transaction/builder/src/transaction_builder.rs

```diff
@@ -4129,4 +4130,361 @@ pub mod transaction_builder_tests {
+                .unwrap();
+
+            let result = transaction_builder.set_fee(1235);
+            assert_matches!(
+                result,
+                Err(TxBuilderError::Memo(NewMemoError::FeeAfterChange))
+            );
+        }
+
+        // Change in a different token is not allowed.
+        {
+            let memo_builder = DefragmentationMemoBuilder::default();
+
+            let mut transaction_builder = TransactionBuilder::new(
+                block_version,
+                Amount::new(10, Mob::ID),
+                fog_resolver.clone(),
+                memo_builder,
+            )
+            .unwrap();
+
+            transaction_builder
+                .add_output(Amount::new(100, Mob::ID), &change_address, &mut rng)
+                .unwrap();
+
+            let result = transaction_builder.add_change_output(
+                Amount::new(0, token_id),
+                &change_destination,
+                &mut rng,
+            );
+
+            assert_matches!(
+                result,
+                Err(TxBuilderError::NewTx(NewTxError::Memo(
+                    NewMemoError::MixedTokenIds
+                )))
+            );
+        }
+
+        // Defrag change output should always be 0
+        {
+            let memo_builder = DefragmentationMemoBuilder::default();
+
+            let mut transaction_builder = TransactionBuilder::new(
+                block_version,
+                Amount::new(10, token_id),
+                fog_resolver.clone(),
+                memo_builder,
+            )
+            .unwrap();
+
+            transaction_builder.set_fee(3).unwrap();
+
+            let input_credentials = get_input_credentials(
+                block_version,
+                Amount::new(113, token_id),
+                &sender,
+                &fog_resolver,
+                &mut rng,
+            );
+            transaction_builder.add_input(input_credentials);
+
+            transaction_builder
+                .add_output(Amount::new(113, token_id), &change_address, &mut rng)
+                .unwrap();
+            let result = transaction_builder.add_change_output(
+                Amount::new(43, token_id),
+                &change_destination,
+                &mut rng,
+            );
+
+            assert_matches!(
+                result,
+                Err(TxBuilderError::NewTx(NewTxError::Memo(
+                    NewMemoError::DefragWithChange
+                )))
+            );
+        }
+
+        // Test builds memos with no ID
+        {
+            let memo_builder = DefragmentationMemoBuilder::default();
+
+            let mut transaction_builder = TransactionBuilder::new(
+                block_version,
+                Amount::new(10, token_id),
+                fog_resolver.clone(),
+                memo_builder,
+            )
+            .unwrap();
+
+            transaction_builder.set_fee(3).unwrap();
+
```

Comment:
:wrench: Similar to above fee doesn't seem necessary
```suggestion
```

Reply:
This one is actually necessary. Below, in the first set of asserts, the fee in the main defrag output should match the value set here. The second set of asserts checks that even though the fee was non-zero, the "change" output is built with a fee of zero. Rationale for this behavior is discussed and specified in [MCIP 61](https://github.com/mobilecoinfoundation/mcips/pull/61).